File: wpilibj/src/test/java/edu/wpi/first/wpilibj/shuffleboard/ShuffleboardInstanceTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.shuffleboard;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import edu.wpi.first.networktables.GenericEntry;
import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.networktables.NetworkTableInstance;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class ShuffleboardInstanceTest {
  private NetworkTableInstance m_ntInstance;
  private ShuffleboardInstance m_shuffleboardInstance;

  @BeforeEach
  void setupInstance() {
    m_ntInstance = NetworkTableInstance.create();
    m_shuffleboardInstance = new ShuffleboardInstance(m_ntInstance);
  }

  @AfterEach
  void tearDownInstance() {
    m_ntInstance.close();
  }

  @Test
  void testPathFluent() {
    GenericEntry entry =
        m_shuffleboardInstance
            .getTab("Tab Title")
            .getLayout("Layout Title", "List Layout")
            .add("Data", "string")
            .withWidget("Text View")
            .getEntry();

    assertAll(
        () -> assertEquals("string", entry.getString(null), "Wrong entry value"),
        () ->
            assertEquals(
                "/Shuffleboard/Tab Title/Layout Title/Data",
                entry.getTopic().getName(),
                "Entry path generated incorrectly"));
  }

  @Test
  void testNestedLayoutsFluent() {
    GenericEntry entry =
        m_shuffleboardInstance
            .getTab("Tab")
            .getLayout("First", "List")
            .getLayout("Second", "List")
            .getLayout("Third", "List")
            .getLayout("Fourth", "List")
            .add("Value", "string")
            .getEntry();

    assertAll(
        () -> assertEquals("string", entry.getString(null), "Wrong entry value"),
        () ->
            assertEquals(
                "/Shuffleboard/Tab/First/Second/Third/Fourth/Value",
                entry.getTopic().getName(),
                "Entry path generated incorrectly"));
  }

  @Test
  void testNestedLayoutsOop() {
    ShuffleboardTab tab = m_shuffleboardInstance.getTab("Tab");
    ShuffleboardLayout first = tab.getLayout("First", "List");
    ShuffleboardLayout second = first.getLayout("Second", "List");
    ShuffleboardLayout third = second.getLayout("Third", "List");
    ShuffleboardLayout fourth = third.getLayout("Fourth", "List");
    SimpleWidget widget = fourth.add("Value", "string");
    GenericEntry entry = widget.getEntry();

    assertAll(
        () -> assertEquals("string", entry.getString(null), "Wrong entry value"),
        () ->
            assertEquals(
                "/Shuffleboard/Tab/First/Second/Third/Fourth/Value",
                entry.getTopic().getName(),
                "Entry path generated incorrectly"));
  }

  @Test
  void testLayoutTypeIsSet() {
    String layoutType = "Type";
    m_shuffleboardInstance.getTab("Tab").getLayout("Title", layoutType);
    m_shuffleboardInstance.update();
    NetworkTableEntry entry =
        m_ntInstance.getEntry("/Shuffleboard/.metadata/Tab/Title/PreferredComponent");
    assertEquals(layoutType, entry.getString("Not Set"), "Layout type not set");
  }

  @Test
  void testNestedActuatorWidgetsAreDisabled() {
    m_shuffleboardInstance
        .getTab("Tab")
        .getLayout("Title", "Layout")
        .add(new MockActuatorSendable("Actuator"));
    NetworkTableEntry controllableEntry =
        m_ntInstance.getEntry("/Shuffleboard/Tab/Title/Actuator/.controllable");

    m_shuffleboardInstance.update();

    // Note: we use the unsafe `getBoolean()` method because if the value is NOT a boolean, or if it
    // is not present, then something has clearly gone very, very wrong
    boolean controllable = controllableEntry.getValue().getBoolean();

    // Sanity check
    assertTrue(controllable, "The nested actuator widget should be enabled by default");
    m_shuffleboardInstance.disableActuatorWidgets();
    controllable = controllableEntry.getValue().getBoolean();
    assertFalse(controllable, "The nested actuator widget should have been disabled");
  }
}

File: wpilibc/src/test/native/cpp/shuffleboard/ShuffleboardInstanceTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/shuffleboard/ShuffleboardInstance.h"  // NOLINT(build/include_order)

#include <string_view>

#include <gtest/gtest.h>
#include <networktables/NetworkTableInstance.h>
#include <networktables/NetworkTableListener.h>
#include <networktables/StringTopic.h>

#include "shuffleboard/MockActuatorSendable.h"

class NTWrapper {
 public:
  NTWrapper() { inst = nt::NetworkTableInstance::Create(); }

  ~NTWrapper() { nt::NetworkTableInstance::Destroy(inst); }

  nt::NetworkTableInstance inst;
};

TEST(ShuffleboardInstanceTest, PathFluent) {
  NTWrapper ntInst;
  frc::detail::ShuffleboardInstance shuffleboardInst{ntInst.inst};

  auto entry = shuffleboardInst.GetTab("Tab Title")
                   .GetLayout("List", "List Layout")
                   .Add("Data", "string")
                   .WithWidget("Text View")
                   .GetEntry();

  EXPECT_EQ("string", entry->GetString("")) << "Wrong entry value";
  EXPECT_EQ("/Shuffleboard/Tab Title/List/Data", entry->GetTopic().GetName())
      << "Entry path generated incorrectly";
}

TEST(ShuffleboardInstanceTest, NestedLayoutsFluent) {
  NTWrapper ntInst;
  frc::detail::ShuffleboardInstance shuffleboardInst{ntInst.inst};

  auto entry = shuffleboardInst.GetTab("Tab")
                   .GetLayout("First", "List")
                   .GetLayout("Second", "List")
                   .GetLayout("Third", "List")
                   .GetLayout("Fourth", "List")
                   .Add("Value", "string")
                   .GetEntry();

  EXPECT_EQ("string", entry->GetString("")) << "Wrong entry value";
  EXPECT_EQ("/Shuffleboard/Tab/First/Second/Third/Fourth/Value",
            entry->GetTopic().GetName())
      << "Entry path generated incorrectly";
}

TEST(ShuffleboardInstanceTest, NestedLayoutsOop) {
  NTWrapper ntInst;
  frc::detail::ShuffleboardInstance shuffleboardInst{ntInst.inst};

  frc::ShuffleboardTab& tab = shuffleboardInst.GetTab("Tab");
  frc::ShuffleboardLayout& first = tab.GetLayout("First", "List");
  frc::ShuffleboardLayout& second = first.GetLayout("Second", "List");
  frc::ShuffleboardLayout& third = second.GetLayout("Third", "List");
  frc::ShuffleboardLayout& fourth = third.GetLayout("Fourth", "List");
  frc::SimpleWidget& widget = fourth.Add("Value", "string");
  auto entry = widget.GetEntry();

  EXPECT_EQ("string", entry->GetString("")) << "Wrong entry value";
  EXPECT_EQ("/Shuffleboard/Tab/First/Second/Third/Fourth/Value",
            entry->GetTopic().GetName())
      << "Entry path generated incorrectly";
}

TEST(ShuffleboardInstanceTest, LayoutTypeIsSet) {
  NTWrapper ntInst;
  frc::detail::ShuffleboardInstance shuffleboardInst{ntInst.inst};

  std::string_view layoutType = "Type";
  shuffleboardInst.GetTab("Tab").GetLayout("Title", layoutType);
  shuffleboardInst.Update();
  auto entry = ntInst.inst.GetEntry(
      "/Shuffleboard/.metadata/Tab/Title/PreferredComponent");
  EXPECT_EQ(layoutType, entry.GetString("Not Set")) << "Layout type not set";
}

TEST(ShuffleboardInstanceTest, NestedActuatorWidgetsAreDisabled) {
  NTWrapper ntInst;
  frc::detail::ShuffleboardInstance shuffleboardInst{ntInst.inst};

  MockActuatorSendable sendable("Actuator");
  shuffleboardInst.GetTab("Tab").GetLayout("Title", "Layout").Add(sendable);
  auto controllableEntry =
      ntInst.inst.GetEntry("/Shuffleboard/Tab/Title/Actuator/.controllable");
  shuffleboardInst.Update();

  // Note: we use the unsafe `GetBoolean()` method because if the value is NOT
  // a boolean, or if it is not present, then something has clearly gone very,
  // very wrong
  bool controllable = controllableEntry.GetValue().GetBoolean();
  // Sanity check
  EXPECT_TRUE(controllable)
      << "The nested actuator widget should be enabled by default";
  shuffleboardInst.DisableActuatorWidgets();
  controllable = controllableEntry.GetValue().GetBoolean();
  EXPECT_FALSE(controllable)
      << "The nested actuator widget should have been disabled";
}

TEST(ShuffleboardInstanceTest, DuplicateSelectTabs) {
  NTWrapper ntInst;
  frc::detail::ShuffleboardInstance shuffleboardInst{ntInst.inst};
  std::atomic_int counter = 0;
  auto subscriber =
      ntInst.inst.GetStringTopic("/Shuffleboard/.metadata/Selected")
          .Subscribe("", {.keepDuplicates = true});
  ntInst.inst.AddListener(
      subscriber, nt::EventFlags::kValueAll | nt::EventFlags::kImmediate,
      [&counter](auto& event) { counter++; });

  // There shouldn't be anything there
  EXPECT_EQ(0, counter);

  shuffleboardInst.SelectTab("tab1");
  shuffleboardInst.SelectTab("tab1");
  EXPECT_TRUE(ntInst.inst.WaitForListenerQueue(1.0))
      << "Listener queue timed out!";
  EXPECT_EQ(2, counter);
}
